### PR TITLE
implement scripting support

### PR DESCRIPTION
**Pull request by jrassa**
This is something I put together a while back and never finished up.  It was inspired by scripting supporting in [Pegasus](http://pegasus-frontend.org/docs/user-guide/scripting/) but with a few additional events that I thought could be useful.

draft wiki documentation for events can be found [here](https://github.com/jrassa/EmulationStation/wiki/Scripting).

### CODE REVIEW

**Comment by pjft** [comment]
Thanks - and great to see new things from you!

A comment - feature-set specifically: I find it weird that there's a "quit" event but not a "start" event - was it deliberate?

Also, could you elaborate on the kind of use cases you'd see this supporting? Just to grasp the concept/possibilities here.

Thank you, and have a great weekend!

**Comment by jrassa** [comment]
> A comment - feature-set specifically: I find it weird that there's a "quit" event but not a "start" event - was it deliberate?

Not really.  I think part of it was that I based it on the scripting support in Pegasus and it wasn't supported there.  As I was refreshing myself on this, I had the same thought, but figured it could be handled by whatever script is used to initially launch ES.  That being said, I wouldn't have any objection to adding it if there is demand.

> Also, could you elaborate on the kind of use cases you'd see this supporting? Just to grasp the concept/possibilities here.

Obviously uses cases would vary based on the events.  The initial use case that pushed me to implement this was for the `theme-changed` event.  Some themes provide launching images that match theme and there is even a script in the forums to generate custom launching images from a theme.  The thought was that the `theme-changed` event could be used to change the launching images based on the selected theme.

The others have potential use cases as well.  Some users run audio in the background to make up for the lack of functionality in ES.  The `game-start` and `game-end` events could be used to pause/play music.  `sleep` and `wake` could be used to turn off the monitor.  'game-end` could also be used to launch a script that syncs game saves.

Some of the other events could potentially be used to extract some of the retropie specific functionality within ES, such as handling restarts and syncing ES controller config with Retroarch controller configs.  I'm not necessarily advocating that we should do that, but it might make sense to.  At a minimum it would make it easier to implement similar functionality on other platforms like Windows.

**Comment by pjft** [comment]
Got it, thanks for clarifying. The theme ones I was completely clueless about, but your explanation made sense. Same goes for the game start/end.

My thoughts on the "start" event are similar, though I suppose the same could be said for handling the "quit" event. But I agree with you on measuring demand on those.

**Comment by cmitu** [comment]
I have one suggestion for the location of the scripts - instead of using a single location (`~/.emulationstation/scripts`), abstract it like it's done for resources https://github.com/RetroPie/EmulationStation/blob/e45458dee61037537c07b3bf05b3415955eb073e/es-core/src/resources/ResourceManager.cpp#L30-L45 
This would allow scripts to be distributed with the ES package by the packager (`<install_location>/scripts`) or overwritten by the user (`~/.emulationstation/scripts`).

**Comment by jrassa** [comment]
@cmitu That is a good idea.  However, in this case, I don't think we want one location to override the other.  I think they should be additive.  First execute any scripts in the install location (i.e. scripts distributed with ES), then execute any scripts added by the user in the home directory.

I pushed an update that implements this.

**Comment by cmitu** [comment]
>  First execute any scripts in the install location (i.e. scripts distributed with ES), then execute any scripts added by the user in the home directory.

Yes, that makes sense and provides additional options. I'm wondering if there's be an advantage to sort the scripts to make the execution order more predictable (though this could be tackled in another PR or modification).

**Comment by shantigilbert** [comment]
I love this idea, I used it a lot in Pegasus, so I am excited about this PR! 
I do have a small request tho, could there be an option to reboot or shutdown without calling the quit script? or a way to identify when the quit script has been called from any of this 2 events? Maybe an argument? 

```
Scripting::fireEvent("quit", "restart");
Scripting::fireEvent("quit", "shutdown");
```

The reason for my request is that I run ES from Kodi (as an add-on) and the way I used Pegasus was that I had a quit script that would do certain things to run Kodi after I quit, but this meant that anytime reboot or shutdown Kodi was being called before (or any other programs you have on quit for that matter). 

Thanks!

**Comment by cmitu** [comment]
@shantigilbert I don't understand this request - does Kodi exits when it's starting Emulationstation so you have to start it manually ? Scripting would be something that's internal to ES, so calling scripts (from events) after ES exits wouldn't work, since ES would have to wait for the script to finish, which kind of defeats the purpose.
For your use case,  wouldn't it be easier to just start Kodi in a loop (from the autostart script) and just Quit Emulationstation instead of using Restart/Shutdown ?

**Comment by shantigilbert** [comment]
Yes this is running on an Arm device so freeing as much resources as possible is best, and I understand that es would wait for a program to exit, but since kodi is ran as a service it won't hold it up. That is the way I was using it in Pegasus, but it would bee so much easier if this was implemented. It's not a big deal, I can always patch it later if I need to. Thanks!

**Comment by jrassa** [comment]
> I'm wondering if it would be an advantage to sort the scripts to make the execution order more predictable (though this could be tackled in another PR or modification).

I considered that, but figured if the users needs to control the order they are executed, they could create a single master script that calls the subscripts stored elsewhere

File: es-app/src/guis/GuiMenu.cpp [hunk 1]
```diff
@@ -500,6 +503,8 @@ void GuiMenu::openQuitMenu()
 	row.makeAcceptInputHandler([window] {
 		window->pushGui(new GuiMsgBox(window, "REALLY RESTART?", "YES",
 			[] {
+			Scripting::fireEvent("quit");
+			Scripting::fireEvent("restart");
```
Comment (shantigilbert) [inline es-app/src/guis/GuiMenu.cpp]: on your wiki you have 

```reboot | on system reboot (also calls quit first)```

shouldn't this line be ``` Scripting::fireEvent("reboot"); ```

Comment (jrassa) [reply es-app/src/guis/GuiMenu.cpp]: fixed

**Comment by jrassa** [comment]
> Maybe an argument?
> 
> ```
> Scripting::fireEvent("quit", "reboot");
> Scripting::fireEvent("quit", "shutdown");
> ```

I added this.  If reboot or shutdown, the quit script will be passed a single argument indicating which one.

**Comment by shantigilbert** [comment]
@jrassa on your wiki you mention screensaver-start and screensaver-stop but those do not fire when the screen saver starts nor ends, and I don't  see ``` Scripting::fireEvent("screensaver-start"); ``` nor ```Scripting::fireEvent("screensaver-stop");``` did I miss something? or where these not implemented for some reason ?

**Comment by jrassa** [comment]
@shantigilbert those events are being added as part of #550

**Pull request merged**